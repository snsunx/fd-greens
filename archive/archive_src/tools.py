from qiskit.quantum_info import Pauli, SparsePauliOp
from qiskit.opflow.primitive_ops import PauliSumOp
from itertools import combinations
import numpy as np

def get_number_state_indices(n_orb, n_elec, anc=''):
    inds = []
    for tup in combinations(range(n_orb), n_elec):
        bin_list = ['1' if (n_orb - 1 - i) in tup else '0' for i in range(n_orb)]
        bin_str = anc + ''.join(bin_list)
        inds.append(int(bin_str, 2))
    inds.sort()
    return inds

def number_state_eigensolver(hamiltonian_mat, n_elec):
    # TODO: If hamiltonian is not a sparse array, convert it to sparse array
    n_orb = int(np.log2(hamiltonian_mat.shape[0]))
    inds = get_number_state_indices(n_orb, n_elec)
    hamiltonian_subspace = hamiltonian_mat[inds][:, inds].toarray()
    eigvals, eigvecs = np.linalg.eigh(hamiltonian_subspace)
    return eigvals, eigvecs

# XXX: This function is deprecated.
def openfermion_to_qiskit_operator(opf_qubit_op):
    """Converts an openfermion QubitOperator to a qiskit PauliSumOp."""
    table = []
    coeffs = []
    n_qubits = 0
    for key in opf_qubit_op.terms:
        if key == ():
            continue
        num = max([t[0] for t in key])
        if num > n_qubits:
            n_qubits = num
    n_qubits += 1

    for key, val in opf_qubit_op.terms.items():
        coeffs.append(val)
        label = ['I'] * n_qubits
        for i, s in key:
            label[i] = s
        label = ''.join(label)
        pauli = Pauli(label[::-1]) # because Qiskit qubit order is reversed
        mask = list(pauli.x) + list(pauli.z)
        table.append(mask)
    primitive = SparsePauliOp(table, coeffs)
    qk_qubit_op = PauliSumOp(primitive)
    return qk_qubit_op
