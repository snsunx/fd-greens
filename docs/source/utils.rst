utils package
=============

Submodules
----------

utils.circuit\_utils module
---------------------------

.. automodule:: utils.circuit_utils
   :members:
   :undoc-members:
   :show-inheritance:

utils.general\_utils module
---------------------------

.. automodule:: utils.general_utils
   :members:
   :undoc-members:
   :show-inheritance:

utils.helpers module
--------------------

.. automodule:: utils.helpers
   :members:
   :undoc-members:
   :show-inheritance:

utils.io\_utils module
----------------------

.. automodule:: utils.io_utils
   :members:
   :undoc-members:
   :show-inheritance:

utils.plot\_utils module
------------------------

.. automodule:: utils.plot_utils
   :members:
   :undoc-members:
   :show-inheritance:

utils.process\_utils module
---------------------------

.. automodule:: utils.process_utils
   :members:
   :undoc-members:
   :show-inheritance:

utils.tomography\_utils module
------------------------------

.. automodule:: utils.tomography_utils
   :members:
   :undoc-members:
   :show-inheritance:

Module contents
---------------

.. automodule:: utils
   :members:
   :undoc-members:
   :show-inheritance:
