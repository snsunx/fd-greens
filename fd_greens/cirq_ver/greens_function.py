"""
===================================================
Green's Function (:mod:`fd_greens.greens_function`)
===================================================
"""

import os
from typing import Sequence, Optional
from itertools import product

import h5py
import numpy as np

from .molecular_hamiltonian import MolecularHamiltonian
from .qubit_indices import QubitIndices
from .parameters import REVERSE_QUBIT_ORDER, ErrorMitigationParameters, get_method_indices_pairs
from .general_utils import project_density_matrix, purify_density_matrix, reverse_qubit_order, two_qubit_state_tomography

np.set_printoptions(precision=6)

class GreensFunction:
    """Frequency-domain Green's function."""

    def __init__(
        self,
        hamiltonian: MolecularHamiltonian,
        fname: str = "lih",
        suffix: str = "",
        spin: str = "d",
        method: str = "exact",
        verbose: bool = True,
    ) -> None:
        """Initializes a ``GreensFunction`` object.
        
        Args:
            hamiltonian: The molecular Hamiltonian.
            fname: The HDF5 file name.
            suffix: The suffix for a specific experimental run.
            spin: Spin of the second quantized operators. Either ``'u'`` or ``'d'``.
            method: The method used for calculating the transition amplitudes. Either ``'exact'`` or ``'tomo'``.
            verbose: Whether to print out transition amplitude values.
        """
        assert spin in ['u', 'd']
        assert method in ['exact', 'tomo']
        
        # Input attributes.
        self.hamiltonian = hamiltonian
        self.fname = fname
        self.h5fname = fname + ".h5"
        self.suffix = suffix
        self.spin = spin
        self.method = method
        self.verbose = verbose

        self.parameters = ErrorMitigationParameters()
        self.parameters.write(fname)

        # Load energies and state vectors from HDF5 file.
        with h5py.File(self.h5fname, "r") as h5file:
            self.energies = {"gs": h5file["gs/energy"][()], 
                             "e": h5file["es/energies_e"][:],
                             "h": h5file["es/energies_h"][:]}
            self.state_vectors = {"e": h5file["es/states_e"][:], 
                                  "h": h5file["es/states_h"][:]}

        # Derived attributes.
        method_indices_pairs = get_method_indices_pairs(spin)
        self.n_states = {subscript: self.state_vectors[subscript].shape[1] for subscript in ['e', 'h']}
        self.n_orbitals = len(self.hamiltonian.active_indices)
        self.n_system_qubits = 2 * self.n_orbitals - method_indices_pairs.n_tapered

        self.qubit_indices_dict = QubitIndices.get_eh_qubit_indices_dict(
            2 * self.n_orbitals, spin, method_indices_pairs)

        # Initialize array quantities B, D and G.
        self.B = {subscript: np.zeros((self.n_orbitals, self.n_orbitals, self.n_states[subscript]), dtype=complex)
                  for subscript in ["e", "h"]}
        self.D = {subscript: np.zeros((self.n_orbitals, self.n_orbitals, self.n_states[subscript[0]]), dtype=complex)
                  for subscript in ["ep", "em", "hp", "hm"]}
        self.G = {subscript: np.zeros((self.n_orbitals, self.n_orbitals), dtype=complex)
                  for subscript in ["e", "h"]}
        
        self._process_diagonal_results()
        self._process_off_diagonal_results()

    def _process_diagonal_results(self) -> None:
        """Processes diagonal transition amplitude results."""
        h5file = h5py.File(self.h5fname, "r+")

        for m in range(self.n_orbitals):
            circuit_label = f"circ{m}{self.spin}"

            for subscript in ['e', 'h']:
                qubit_indices = self.qubit_indices_dict[subscript]
                state_vectors_exact = self.state_vectors[subscript]

                if self.method == 'exact':
                    state_vector = h5file[f'{circuit_label}/transpiled'].attrs[f'psi{self.suffix}']
                    if REVERSE_QUBIT_ORDER:
                        state_vector = reverse_qubit_order(state_vector)
                    state_vector = qubit_indices(state_vector)

                    self.B[subscript][m, m] = np.abs(state_vectors_exact.conj().T @ state_vector) ** 2

                elif self.method == 'tomo':
                    tomography_labels = [''.join(x) for x in product('xyz', repeat=2)]

                    array_all = []
                    for tomography_label in tomography_labels:
                        array_raw = h5file[f"{circuit_label}/{tomography_label}"].attrs[f"counts{self.suffix}"]
                        repetitions = np.sum(array_raw)
                        ancilla_index = int(qubit_indices.ancilla.str[0], 2)
                        if REVERSE_QUBIT_ORDER:
                            array_raw = reverse_qubit_order(array_raw)
                            array_label = array_raw[ancilla_index :: 2] / repetitions
                        else:
                            array_label = array_raw[ancilla_index * 4 : (ancilla_index + 1) * 4] / repetitions
                        array_all += list(array_label)

                    density_matrix = two_qubit_state_tomography(array_all)
                    density_matrix = qubit_indices.system(density_matrix)

                    trace = np.trace(density_matrix)
                    density_matrix /= trace
                    if self.parameters.PROJECT_DENSITY_MATRICES:
                        density_matrix = project_density_matrix(density_matrix)
                    if self.parameters.PURIFY_DENSITY_MATRICES:
                        density_matrix = purify_density_matrix(density_matrix)

                    self.B[subscript][m, m] = [
                        trace * (state_vectors_exact[:, i].conj() @ density_matrix @ state_vectors_exact[:, i]).real
                        for i in range(self.n_states[subscript])]

                if self.verbose:
                    print(f"B[{subscript}][{m}, {m}] = {self.B[subscript][m, m]}")
                    
        h5file.close()

    def _process_off_diagonal_results(self) -> None:
        """Processes off-diagonal transition amplitude results."""
        h5file = h5py.File(self.h5fname, "r+")

        for m in range(self.n_orbitals):
            for n in range(m + 1, self.n_orbitals):
                circuit_label = f"circ{m}{n}{self.spin}"

                for subscript in ['ep', 'em', 'hp', 'hm']:
                    qubit_indices = self.qubit_indices_dict[subscript]
                    state_vectors_exact = self.state_vectors[subscript[0]]

                    if self.method == 'exact':
                        state_vector = h5file[f'{circuit_label}/transpiled'].attrs[f"psi{self.suffix}"]
                        if REVERSE_QUBIT_ORDER:
                            state_vector = reverse_qubit_order(state_vector)
                        state_vector = qubit_indices(state_vector)

                        self.D[subscript][m, n] = self.D[subscript][n, m] = \
                            np.abs(state_vectors_exact.conj().T @ state_vector) ** 2

                    elif self.method == 'tomo':
                        tomography_labels = [''.join(x) for x in product('xyz', repeat=2)]

                        array_all = []
                        for tomography_label in tomography_labels:
                            array_raw = h5file[f"{circuit_label}/{tomography_label}"].attrs[f"counts{self.suffix}"]
                            repetitions = np.sum(array_raw)
                            ancilla_index = int(qubit_indices.ancilla.str[0], 2)
                            if REVERSE_QUBIT_ORDER:
                                array_raw = reverse_qubit_order(array_raw)
                                array_label = array_raw[ancilla_index :: 4]  / repetitions
                            else:
                                array_label  = array_raw[ancilla_index * 4:(ancilla_index + 1) * 4] / repetitions
                            array_all += list(array_label)

                        density_matrix = two_qubit_state_tomography(array_all)
                        density_matrix = qubit_indices.system(density_matrix)

                        trace = np.trace(density_matrix)
                        density_matrix /= trace
                        if self.parameters.PROJECT_DENSITY_MATRICES:
                            density_matrix = project_density_matrix(density_matrix)
                        if self.parameters.PURIFY_DENSITY_MATRICES:
                            density_matrix = purify_density_matrix(density_matrix)
                        
                        self.D[subscript][m, n] = self.D[subscript][n, m] = [
                            trace * (state_vectors_exact[:, i].conj() @ density_matrix @ state_vectors_exact[:, i]).real
                            for i in range(self.n_states[subscript[0]])]

                    if self.verbose:
                        print(f"D[{subscript}][{m}, {n}] =", self.D[subscript][m, n])

        # Unpack D values to B values according to Eq. (18) of Kosugi and Matsushita 2020.
        for m in range(self.n_orbitals):
            for n in range(m + 1, self.n_orbitals):
                for subscript in ["e", "h"]:
                    self.B[subscript][m, n] = self.B[subscript][n, m] = \
                        np.exp(-1j * np.pi / 4) * (self.D[subscript + "p"][m, n] - self.D[subscript + "m"][m, n]) \
                        + np.exp(1j * np.pi / 4) * (self.D[subscript + "p"][n, m] - self.D[subscript + "m"][n, m])
                    if self.verbose:
                        print(f"B[{subscript}][{m}, {n}] =", self.B[subscript][m, n])

        h5file.close()

    def mean_field_greens_function(self, omega: float, eta: float = 0.0) -> np.ndarray:
        """Returns the mean-field Green's function.
        
        Args:
            omega: The frequency at which the mean-field Green's function is calculated.
            eta: The broadening factor.

        Returns:
            G0: The mean-field Green's function.
        """
        orbital_energies = self.hamiltonian.orbital_energies[self.hamiltonian.active_indices]

        G0 = np.zeros((self.n_orbitals, self.n_orbitals), dtype=complex)
        for i in range(self.n_orbitals):
            G0[i, i] = 1 / (omega + 1j * eta - orbital_energies[i])
        return G0

    def greens_function(self, omega: float, eta: float = 0.0) -> np.ndarray:
        """Returns the the Green's function at given frequency and broadening.

        Args:
            omega: The frequency at which the Green's function is calculated.
            eta: The broadening factor.
        
        Returns:
            G: The Green's function.
        """
        for m in range(self.n_orbitals):
            for n in range(self.n_orbitals):
                # 2 is for both up and down spins.
                self.G["e"][m, n] = 2 * np.sum(self.B["e"][m, n]
                    / (omega + 1j * eta + self.energies["gs"] - self.energies["e"]))
                self.G["h"][m, n] = 2 * np.sum(self.B["h"][m, n]
                    / (omega + 1j * eta - self.energies["gs"] + self.energies["h"]))
        G = self.G["e"] + self.G["h"]
        return G

    def spectral_function(
        self, omegas: Sequence[float], eta: float = 0.0, save_data: bool = True
    ) -> Optional[np.ndarray]:
        """Returns the spectral function at given frequencies.

        Args:
            omegas: The frequencies at which the spectral function is calculated.
            eta: The broadening factor.
            save_data: Whether to save the spectral function to file.
        
        Returns:
            As: The spectral function numpy array.
        """
        As = []
        for omega in omegas:
            G = self.greens_function(omega, eta)
            A = -1 / np.pi * np.imag(np.trace(G))
            As.append(A)
        As = np.array(As)

        if save_data:
            if not os.path.exists("data"):
                os.makedirs("data")
            np.savetxt(f"data/{self.fname}{self.suffix}_A.dat", np.vstack((omegas, As)).T)
        else:
            return As

    def self_energy(
        self, omegas: Sequence[float], eta: float = 0.0, save_data: bool = True
    ) -> Optional[np.ndarray]:
        """Returns the trace of self-energy at given frequencies.

        Args:
            omegas: The frequencies at which the self-energy is calculated.
            eta: The broadening factor.
            save_data: Whether to save the self-energy to file.

        Returns:
            TrSigmas: Trace of the self-energy.
        """
        TrSigmas = []
        for omega in omegas:
            G0 = self.mean_field_greens_function(omega, eta)
            G = self.greens_function(omega, eta)

            Sigma = np.linalg.pinv(G0) - np.linalg.pinv(G)
            TrSigmas.append(np.trace(Sigma))
        TrSigmas = np.array(TrSigmas)

        if save_data:
            if not os.path.exists("data"):
                os.makedirs("data")
            np.savetxt(
                f"data/{self.fname}{self.suffix}_TrSigma.dat", 
                np.vstack((omegas, TrSigmas.real, TrSigmas.imag)).T)
        else:
            return TrSigmas
