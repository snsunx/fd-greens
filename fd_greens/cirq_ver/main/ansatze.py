"""
================================================
Ansatz Functions (:mod:`fd_greens.main.ansatze`)
================================================
"""

from typing import Callable, Sequence

import cirq

# from qiskit import QuantumCircuit, QuantumRegister

AnsatzFunction = Callable[[Sequence[float]], cirq.Circuit]


def build_ansatz_gs(params: Sequence[float]) -> cirq.Circuit:
    r"""Constructs an $N$-electron ansatz of the encoded Hamiltonian.
    
    The ansatz is of the form $$(R_y(\theta_2)\otimes R_y(\theta_3)) CZ
    (R_y(\theta_0)\otimes R_y(\theta_1)) |00\rangle$$.

    Args:
        params: A sequence of the parameters.

    Returns:
        ansatz: The ansatz circuit.
    """
    assert len(params) == 6
    qubits = [cirq.LineQubit(i) for i in range(2)]
    ansatz = cirq.Circuit()
    ansatz += [cirq.Ry(rads=params[0])(qubits[0]), cirq.Ry(rads=params[1])(qubits[1])]
    ansatz += [cirq.CZ(qubits[0], qubits[1])]
    ansatz += [cirq.Ry(rads=params[2])(qubits[0]), cirq.Ry(rads=params[3])(qubits[1])]
    ansatz += [cirq.CZ(qubits[0], qubits[1])]
    ansatz += [cirq.Ry(rads=params[4])(qubits[0]), cirq.Ry(rads=params[5])(qubits[1])]
    return ansatz


def build_ansatz_e(params: Sequence[float]) -> cirq.Circuit:
    r"""Constructs the ansatz for ($N$+1)-electron states of the encoded Hamiltonian.
    The ansatz is of the form :math:`(I \otimes R_y(\theta))|10\rangle`.

    Args:
        params: A sequence of the parameters.

    Returns:
        ansatz: The ansatz circuit.
    """
    assert len(params) == 1
    qubits = [cirq.LineQubits(i) for i in range(2)]
    ansatz = cirq.Circuit()
    ansatz += [cirq.X(qubits[0])]
    ansatz += [cirq.Ry(rads=params[0])(qubits[1])]
    return ansatz


def build_ansatz_h(params: Sequence[float]) -> cirq.Circuit:
    r"""Constructs the ansatz for ($N$-1)-electron states of the encoded Hamiltonian. 
    The ansatz is of the form :math:`(I \otimes R_y(\theta))|00\rangle`.
    
    Args:
        params: A sequence of the parameters.

    Returns:
        ansatz: The ansatz circuit.
    """
    assert len(params) == 1
    qubits = [cirq.LineQubit(i) for i in range(2)]
    ansatz = cirq.Circuit()
    ansatz += [cirq.Ry(rads=params[0])(qubits[1])]
    return ansatz
