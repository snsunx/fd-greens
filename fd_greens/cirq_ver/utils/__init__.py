from .io_utils import *
from .general_utils import *
from .tomography_utils import *
from .circuit_utils import *
from .helpers import *
from .plot_utils import *
