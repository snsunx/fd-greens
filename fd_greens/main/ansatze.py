"""
================================================
Ansatz Functions (:mod:`fd_greens.main.ansatze`)
================================================
"""

from typing import Callable, Sequence
from qiskit import QuantumCircuit, QuantumRegister

AnsatzFunction = Callable[[Sequence[float]], QuantumCircuit]


def build_ansatz_gs(params: Sequence[float]) -> QuantumCircuit:
    r"""Constructs an N-electron ansatz of the encoded Hamiltonian. The ansatz is of
    the form :math:`(R_y(\theta_2)\otimes R_y(\theta_3)) CZ (R_y(\theta_0)\otimes R_y(\theta_1))
    |00\rangle`.

    Args:
        params: A sequence of the parameters.

    Returns:
        ansatz: The ansatz quantum circuit.
    """
    assert len(params) == 6
    qreg = QuantumRegister(2, name="q")
    ansatz = QuantumCircuit(qreg)
    ansatz.ry(params[0], 0)
    ansatz.ry(params[1], 1)
    ansatz.cz(0, 1)
    ansatz.ry(params[2], 0)
    ansatz.ry(params[3], 1)
    ansatz.cz(0, 1)
    ansatz.ry(params[4], 0)
    ansatz.ry(params[5], 0)
    return ansatz


def build_ansatz_e(params: Sequence[float]) -> QuantumCircuit:
    r"""Constructs the ansatz for (N+1)-electron states of the encoded Hamiltonian.
    The ansatz is of the form :math:`(I \otimes R_y(\theta))|10\rangle`.

    Args:
        params: A sequence of the parameters.

    Returns:
        ansatz: The ansatz quantum circuit.
    """
    assert len(params) == 1
    qreg = QuantumRegister(2, name="q")
    ansatz = QuantumCircuit(qreg)
    ansatz.x(0)
    ansatz.ry(params[0], 1)
    return ansatz


def build_ansatz_h(params: Sequence[float]) -> QuantumCircuit:
    r"""Constructs the ansatz for (N-1)-electron states of the encoded Hamiltonian. 
    The ansatz is of the form :math:`(I \otimes R_y(\theta))|00\rangle`.
    
    Args:
        params: A sequence of the parameters.

    Returns:
        ansatz: The ansatz quantum circuit.
    """
    assert len(params) == 1
    qreg = QuantumRegister(2, name="q")
    ansatz = QuantumCircuit(qreg)
    ansatz.ry(params[0], 1)
    return ansatz
