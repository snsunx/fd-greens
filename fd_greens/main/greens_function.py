"""
========================================================
Green's Function (:mod:`fd_greens.main.greens_function`)
========================================================
"""

import os
from typing import Union, Sequence, Optional

import h5py
import numpy as np


class GreensFunction:
    """A class to calculate frequency-domain Green's function."""

    def __init__(self, h5fname: str = "lih", suffix: str = "") -> None:
        """Initializes a GreensFunction object.
        
        Args:
            h5fname: The HDF5 file name.
            suffix: The suffix for a specific experimental run.
        """
        self.datfname = h5fname + suffix

        h5file = h5py.File(h5fname + ".h5", "r")
        self.energy_gs = h5file["gs/energy"]
        self.energies_e = h5file["es/energies_e"]
        self.energies_h = h5file["es/energies_h"]
        self.B_e = h5file[f"amp/B_e{suffix}"]
        self.B_h = h5file[f"amp/B_h{suffix}"]
        self.n_orb = self.B_e.shape[0]
        self.e_orb = h5file[f"amp/e_orb{suffix}"]

    @property
    def density_matrix(self) -> np.ndarray:
        """The density matrix obtained from the transition amplitudes."""
        rho = np.sum(self.B_h, axis=2)
        return rho

    def greens_function(self, omega: float, eta: float = 0.0) -> np.ndarray:
        """Returns the the Green's function at frequency omega.

        Args:
            omega: The frequency at which the Green's function is calculated.
            eta: The imaginary part, i.e. broadening factor.
        
        Returns:
            G: The Green's function numpy array.
        """
        # Green's function arrays
        G_e = np.zeros((self.n_orb, self.n_orb), dtype=complex)
        G_h = np.zeros((self.n_orb, self.n_orb), dtype=complex)

        for m in range(self.n_orb):
            for n in range(self.n_orb):
                G_e[m, n] = 2 * np.sum(
                    self.B_e[m, n]
                    / (omega + 1j * eta + self.energy_gs - self.energies_e)
                )
                G_h[m, n] = 2 * np.sum(
                    self.B_h[m, n]
                    / (omega + 1j * eta - self.energy_gs + self.energies_h)
                )
        G = G_e + G_h
        return G

    def spectral_function(
        self, omegas: Sequence[float], eta: float = 0.0, save_data: bool = True
    ) -> Optional[np.ndarray]:
        """Returns the spectral function at certain frequencies.

        Args:
            omegas: The frequencies at which the spectral function is calculated.
            eta: The imaginary part, i.e. broadening factor.
            save_data: Whether to save the spectral function to file.
        
        Returns:
            As: The spectral function numpy array.
        """
        As = []
        for omega in omegas:
            G = self.greens_function(omega, eta)
            A = -1 / np.pi * np.imag(np.trace(G))
            As.append(A)
        As = np.array(As)

        if save_data:
            if not os.path.exists('data'):
                os.makedirs('data')
            np.savetxt("data/" + self.datfname + "_A.dat", np.vstack((omegas, As)).T)
        else:
            return As

    def self_energy(
        self, omegas: Sequence[float], eta: float = 0.0, save_data: bool = True
    ) -> Optional[np.ndarray]:
        """Returns the trace of self-energy at frequency omega.

        Args:
            omegas: The frequencies at which the spectral function is calculated.
            eta: The imaginary part, i.e. broadening factor.
            save_data: Whether to save the spectral function to file.

        Returns:
            TrSigmas: Trace of the self-energy.
        """
        TrSigmas = []
        for omega in omegas:
            G = self.greens_function(omega + 1j * eta)
            G_HF = np.zeros((self.n_orb, self.n_orb), dtype=complex)
            for i in range(self.n_orb):
                G_HF[i, i] = 1 / (omega - self.e_orb[i, i])

            Sigma = np.linalg.pinv(G_HF) - np.linalg.pinv(G)
            TrSigmas.append(np.trace(Sigma))
        TrSigmas = np.array(TrSigmas)

        if save_data:
            if not os.path.exists('data'):
                os.makedirs('data')
            np.savetxt(
                "data/" + self.datfname + "_TrS.dat",
                np.vstack((omegas, TrSigmas.real, TrSigmas.imag)).T,
            )
        else:
            return TrSigmas
