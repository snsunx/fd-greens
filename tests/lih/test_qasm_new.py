"""Calculates the spectral function on the active-space LiH Hamilonian 
using the classmethods."""

import sys
sys.path.append('../../src/')
import numpy as np
from qiskit import *
from ansatze import build_ne2_ansatz
from hamiltonians import MolecularHamiltonian
from qiskit.utils import QuantumInstance
from greens_function import GreensFunction
from tools import get_quantum_instance
from constants import HARTREE_TO_EV

# User-defined parameters.
bond_length = 1.6
cache_read = True
cache_write = False

ansatz = build_ne2_ansatz(4)
hamiltonian = MolecularHamiltonian(
    [['Li', (0, 0, 0)], ['H', (0, 0, bond_length)]], 'sto3g', 
    occupied_inds=[0], active_inds=[1, 2])

q_instance_sv = QuantumInstance(Aer.get_backend('statevector_simulator'))
q_instance_qasm = QuantumInstance(Aer.get_backend('qasm_simulator'), shots=8192)
q_instance_noisy = get_quantum_instance(
    Aer.get_backend('qasm_simulator'), shots=8192, 
    noise_model_name='ibmq_jakarta')

# Statevector simulator calculation
print("========== Starts statevector simulation ==========")
gf_sv = GreensFunction(ansatz.copy(), hamiltonian, q_instance=q_instance_sv)
gf_sv.run(cache_read=cache_read, cache_write=cache_write)
#print(gf_sv.states_arr)
exit()

# QASM simulator calculation of h states
print("============ Starts calculating h states ==========")
gf_h = GreensFunction.initialize_eh(gf_sv, 'h', q_instance=q_instance_sv)
gf_h.run(compute_energies=False, cache_read=cache_read, cache_write=cache_write)
print(gf_h.states_arr)

# QASM simulator calculation of e states
print("========== Starts calculating e states ==========")
gf_e = GreensFunction.initialize_eh(gf_sv, 'e', q_instance=q_instance_sv)
gf_e.run(compute_energies=False, cache_read=cache_read, cache_write=cache_write)
print(gf_e.states_arr)

# Combining h states and e states results
gf_final = GreensFunction.initialize_final(
    gf_sv, gf_e, gf_h, q_instance=q_instance_noisy)

omegas = np.arange(-30, 34, 0.1)
A_list = []
for omega in omegas:
    A = gf_final.compute_spectral_function(omega + 0.02j * HARTREE_TO_EV)
    A_list.append(A)
np.savetxt('A_noisy_cached2.dat', np.vstack((omegas, A_list)).T)
