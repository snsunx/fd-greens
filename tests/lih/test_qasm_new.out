========== Starts statevector simulation ==========
Start calculating the ground state using VQE
Finish calculating the ground state using VQE
Ground state energy = -213.939 eV
Start calculating (NÂ±1)-electron states
eigenstates_e
 [[ 0.      +0.j  0.027975+0.j  0.      +0.j -0.999609+0.j]
 [ 0.027975+0.j  0.      +0.j -0.999609+0.j  0.      +0.j]
 [ 0.      +0.j  0.999609+0.j  0.      +0.j  0.027975+0.j]
 [ 0.999609+0.j  0.      +0.j  0.027975+0.j  0.      +0.j]]
eigenstates_h
 [[ 0.      +0.j  0.1143  +0.j  0.      +0.j -0.993446+0.j]
 [-0.1143  +0.j  0.      +0.j  0.993446+0.j  0.      +0.j]
 [ 0.      +0.j -0.993446+0.j  0.      +0.j -0.1143  +0.j]
 [ 0.993446+0.j  0.      +0.j  0.1143  +0.j  0.      +0.j]]
Start calculating diagonal transition amplitudes
m = 0
Index of largest element is 00100
--------------------------------------------------------------------------------
[ 1.910397e-35-1.712021e-34j -2.020530e-02-1.053413e-17j
  2.013550e-34-7.373282e-34j -6.808643e-03-1.179412e-16j]
--------------------------------------------------------------------------------
@@@@@ probs_e = [5.433502e-05 0.000000e+00 4.002769e-04 0.000000e+00]
@@@@@ probs_h = [0.       0.984901 0.       0.014644]
m = 1
Index of largest element is 00010
--------------------------------------------------------------------------------
[-0.020205-1.053413e-17j  0.      +0.000000e+00j -0.006786-6.142380e-18j
  0.      +0.000000e+00j]
--------------------------------------------------------------------------------
@@@@@ probs_e = [0.000000e+00 5.400803e-05 0.000000e+00 4.003018e-04]
@@@@@ probs_h = [0.984896 0.       0.01465  0.      ]
m = 2
Index of largest element is 01111
--------------------------------------------------------------------------------
[-1.994366e-35-1.843321e-34j -6.786421e-03-6.142380e-18j
 -2.591163e-34+3.356054e-34j  9.997496e-01+3.144597e-16j]
--------------------------------------------------------------------------------
@@@@@ probs_e = [0.998338 0.       0.001208 0.      ]
@@@@@ probs_h = [0.000000e+00 8.232831e-05 0.000000e+00 3.722836e-04]
m = 3
Index of largest element is 10111
--------------------------------------------------------------------------------
[-0.006809-1.179412e-16j  0.      +0.000000e+00j  0.99975 +3.144597e-16j
  0.      +0.000000e+00j]
--------------------------------------------------------------------------------
@@@@@ probs_e = [0.       0.998336 0.       0.001209]
@@@@@ probs_h = [8.192818e-05 0.000000e+00 3.723817e-04 0.000000e+00]
Finish calculating diagonal transition amplitudes
Start calculating off-diagonal transition amplitudes
B_e

m = 0, m' = 0, lambda = 2: 0.000400+0.000000j
m = 0, m' = 2, lambda = 0: -0.007365+0.000000j
m = 0, m' = 2, lambda = 2: 0.000695+0.000000j
m = 2, m' = 0, lambda = 0: -0.007365-0.000000j
m = 2, m' = 0, lambda = 2: 0.000695-0.000000j
m = 2, m' = 2, lambda = 0: 0.998338+0.000000j
m = 2, m' = 2, lambda = 2: 0.001208+0.000000j
Finish calculating off-diagonal transition amplitudes
